### PR TITLE
[cling] Inject symbols from libc_nonshared.a

### PR DESCRIPTION
**Pull request by hahnjo**
These symbols may not be found automatically. See also upstream issue https://github.com/llvm/llvm-project/issues/61289. This fixes the test `DynamicLibraryManager/cached_realpath.C`, approach by Lang Hames.

### CODE REVIEW

File: interpreter/cling/lib/Interpreter/IncrementalJIT.cpp [hunk 1]
```diff
@@ -577,6 +581,18 @@ IncrementalJIT::IncrementalJIT(
...
+#if defined(__linux__) && defined(__GLIBC__)
+  // Inject lstat to pass test/DynamicLibraryManager/cached_realpath.C, it may
+  // be in libc_nonshared.a where symbols are not found automatically. See also
+  // upstream issue https://github.com/llvm/llvm-project/issues/61289.
+  SymbolMap LibcNonsharedSymbols;
+  LibcNonsharedSymbols[Jit->mangleAndIntern("lstat")] = JITEvaluatedSymbol(
+      static_cast<JITTargetAddress>(reinterpret_cast<uintptr_t>(&lstat)),
+      JITSymbolFlags::Exported);
+  cantFail(
+      Jit->getMainJITDylib().define(absoluteSymbols(LibcNonsharedSymbols)));
```
Comment (jalopezg-git) [inline interpreter/cling/lib/Interpreter/IncrementalJIT.cpp:593]: I wonder whether we could reuse `IncrementalJIT::addOrReplaceDefinition()` for this, but other than that LGTM!

**Review by jalopezg-git — APPROVED** [review]

File: interpreter/cling/lib/Interpreter/IncrementalJIT.cpp [hunk 2]
```diff
@@ -577,6 +581,18 @@ IncrementalJIT::IncrementalJIT(
 
   char LinkerPrefix = this->m_TM->createDataLayout().getGlobalPrefix();
 
+#if defined(__linux__) && defined(__GLIBC__)
+  // Inject lstat to pass test/DynamicLibraryManager/cached_realpath.C, it may
+  // be in libc_nonshared.a where symbols are not found automatically. See also
+  // upstream issue https://github.com/llvm/llvm-project/issues/61289.
+  SymbolMap LibcNonsharedSymbols;
+  LibcNonsharedSymbols[Jit->mangleAndIntern("lstat")] = JITEvaluatedSymbol(
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/IncrementalJIT.cpp:589]: Can we bring all of the workarounds from [here](https://github.com/llvm/llvm-project/blob/3c6ed559e5274307995586c1499a2c8e4e0276a0/llvm/lib/ExecutionEngine/RuntimeDyld/RTDyldMemoryManager.cpp#L235)? I bet we will have bugs reports on the other functions at the moment we forget how to fix them :)

Comment (hahnjo) [reply interpreter/cling/lib/Interpreter/IncrementalJIT.cpp:589]: I'd rather not since it's really a hack. This should be fixed properly upstream...

Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/IncrementalJIT.cpp:589]: I learn that would take time but I think we should update the related llvm issue in that case and keep it up to date.

**Review by vgvassilev — APPROVED** [review]
Assuming we will update the llvm issue capturing the discussion and our finding, I'd propose to move forward here.

**Comment by jalopezg-git** [comment]
Even better now :+1:!  Many thanks, @hahnjo!

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Comment by Axel-Naumann** [comment]
Does this also fix cling with devtoolset? Or could a similar approach be used? (I forgot the exact library name that provides the extra symbols.)

**Comment by hahnjo** [comment]
> Does this also fix cling with devtoolset?

No.

> Or could a similar approach be used? (I forgot the exact library name that provides the extra symbols.)

Not quite, for that we actually have to load the full static library and stuff it into a `StaticLibraryDefinitionGenerator`.

**Pull request merged**